To services CIDR and Port rules generation on endpoint changes

nebril opened the pull request:
Implements #1668

This only implements matching services by their name and namespace. No labels matching against services yet.

nebril committed:
Add ToService egress rule types

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
Generate ToCIDR and ToPorts entries from endpoints

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
Delete ToCIDR and ToPorts rules based on endpoints

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
Manipulate egress policies on k8s endpoint changes

When a new k8s endpoint is detected by watcher, policy repo populates
egress rules containing ToService rules matching the new endpoint with
ToCIDR and ToPorts rules that correspond to new endpoint.

When a k8s endpoint is deleted, policy repo removes all ToCIDR and
ToPorts rules from egress rules with ToService rule matching deleted
endpoint.

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
Move EP-based policy gen to egress method

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
Populate newly created policies based on endpoints

When k8s watcher sees new Cilium Network Policy being created,
it populates its Egress `ToService` policies with ToCIDR and ToPorts rules based
on k8s endpoints in loadBalancer.

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
Change generated CIDRs to one-address subnets

Removed ToPorts rule generation, left TODOs for when ToCIDR and ToPorts
are compatible.

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
Depopulate egress policies before populating them

This is a no-op for endpoint add, and makes sure that endpoint update
provides good rules for egress policies.

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

nebril committed:
ToServices egress docs and comments

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

aanm commented on pkg/policy/repository.go:
```diff
@@ -369,3 +373,85 @@ func (p *Repository) NumRules() int {
 	return p.revision
 }
+
+// ConvertToK8sServiceToToCIDR traverses all egress rules and matches them against provided serviceInfo. If a matching egress rule is found it is populated with ToCIDR and ToPorts entries based on endpoint object.
```
Can you write the comments within 80 columns?

manalibhutiyani commented on daemon/k8s_watcher.go:
```diff
@@ -938,8 +955,9 @@ func (d *Daemon) addCiliumNetworkPolicy(obj interface{}) {
-	if err == nil {
-		if len(rules) > 0 {
+	if err == nil && len(rules) > 0 {
+		err = rules.GenerateEgressRulesFromEndpoints(d.loadBalancer.K8sEndpoints)
```
Should we be handling err!=nil here with a log.Errorf?

nebril replied:
It's handled here: https://github.com/cilium/cilium/pull/1785/files/eeece5dd13704ced9fe454b8c8a574814ff19cc5#diff-7c38bf553d4d2690add68b2464b58ef0R966

aanm replied:
@nebril no, the `err` on line 959 will overwrite the value of the `err` on line 957. So you need to handle the err on this line.

nebril replied:
@aanm 959 will only overwrite the error if `err == nil` (line 958). Each new err assignment is preceded by `err == nil` check, so only one error will be handled on line 966.

aanm replied:
You're right, disregard my last comment

manalibhutiyani commented on pkg/policy/api/utils.go:
```diff
@@ -97,3 +101,155 @@ func ParseL4Proto(proto string) (L4Proto, error) {
+
+// GenerateToServiceRulesFromEndpoint populates egress rule with ToCIDR rules
+// based on ToServices defined in egress rule and provided endpoint
+func (e *EgressRule) GenerateToServiceRulesFromEndpoint(serviceInfo types.K8sServiceNamespace, endpoint types.K8sServiceEndpoint) error {
```
Would be good if you can format this into a 80 char Len.

manalibhutiyani commented on pkg/policy/api/utils.go:
```diff
@@ -97,3 +101,155 @@ func ParseL4Proto(proto string) (L4Proto, error) {
+	return nil
+}
+
+// generateToCidrFromEndpoint takes an egress rule and populates it with ToCIDR rules based on provided enpoint object
```
Nit enpoint --> endpoint

aanm replied:
Can you comment in 80 columns?

manalibhutiyani commented on pkg/policy/api/utils.go:
```diff
@@ -97,3 +101,155 @@ func ParseL4Proto(proto string) (L4Proto, error) {
+	return nil
+}
+
+// generateToPortsFromEndpoint takes an egress rule and populates it with ToPorts rules based on provided enpoint object
```
Nit enpoint --> endpoint

aanm replied:
80 columns

manalibhutiyani commented on pkg/policy/api/utils.go:
```diff
@@ -97,3 +101,155 @@ func ParseL4Proto(proto string) (L4Proto, error) {
+}
+
+// deleteToPortsFromEndpoint takes an egress rule and removes ToPorts rules matching endpoint
+func deleteToPortsFromEndpoint(egress *EgressRule, endpoint types.K8sServiceEndpoint) error {
```
Not sure if relevant, but is any locking required while deleting /updating endpoint ?

nebril replied:
Locking is done on repository level ATM, https://github.com/cilium/cilium/pull/1785/files#diff-7c38bf553d4d2690add68b2464b58ef0R480

aanm replied:
This function is not used

aanm commented on daemon/k8s_watcher.go:
```diff
@@ -476,6 +476,15 @@ func (d *Daemon) endpointAddFn(obj interface{}) {
+
+	d.policy.Mutex.Lock()
+	defer d.policy.Mutex.Unlock()
+	if err := d.policy.DeleteEndpointGeneratedEgressRules(svcns, *newSvcEP); err != nil {
```
This makes me uncomfortable in the sense you are making policy change on all nodes whenever an endpoint appears on the cluster.

aanm commented on pkg/policy/api/rule.go:
```diff
@@ -235,6 +248,10 @@ type EgressRule struct {
 	ToEntities []Entity `json:"toEntities,omitempty"`
+
+	// ToServices is a list of services to which the endpoint subject
+	// to the rule is allowed to initiate connections.
```
Needs example and if it's an optional field or not.

nebril replied:
Done

aanm commented on pkg/policy/api/utils.go:
```diff
@@ -97,3 +101,155 @@ func ParseL4Proto(proto string) (L4Proto, error) {
+	// This will generate one-address CIDRs consisting of endpoint backend ip
+	mask := net.CIDRMask(128, 128)
+	for ip := range endpoint.BEIPs {
+		epIP := net.ParseIP(ip)
```
make sure the `epIP` is not nil

aanm commented on pkg/policy/api/utils.go:
```diff
@@ -97,3 +101,155 @@ func ParseL4Proto(proto string) (L4Proto, error) {
+		}
+		if !found {
+			cidr := net.IPNet{IP: epIP.Mask(mask), Mask: mask}
+			egress.ToCIDR = append(egress.ToCIDR, CIDR(cidr.String()))
```
Instead of doing this, you could re-write the function to return a slice of cidrs. This prevents to have something in the `egress.ToCIDR` in case something bad happens on line 131 for example.

If you return a slice of cidrs, and an error, you can check for the error and if the error is nil only then apply the slice to the `egress.ToCIDR`

nebril replied:
Makes sense, will do

aanm commented on pkg/policy/api/utils.go:
```diff
@@ -97,3 +101,155 @@ func ParseL4Proto(proto string) (L4Proto, error) {
+}
+
+// generateToPortsFromEndpoint takes an egress rule and populates it with ToPorts rules based on provided enpoint object
+func generateToPortsFromEndpoint(egress *EgressRule, endpoint types.K8sServiceEndpoint) error {
```
This function is not used

aanm commented on pkg/policy/api/utils.go:
```diff
@@ -97,3 +101,155 @@ func ParseL4Proto(proto string) (L4Proto, error) {
+		for index := range rule.Egress {
+			for ns, ep := range endpoints {
+				if err := rule.Egress[index].GenerateToServiceRulesFromEndpoint(ns, *ep); err != nil {
+					return err
```
In case of an error you might already changed some rules in the policy repo. Can you prevent this?

aanm commented on pkg/policy/repository.go:
```diff
@@ -369,3 +370,33 @@ func (p *Repository) NumRules() int {
+// ConvertToK8sServiceToToCIDR traverses all egress rules and matches them against
+// provided serviceInfo. If a matching egress rule is found it is populated
+// with ToCIDR entries based on endpoint object.
+func (p *Repository) ConvertToK8sServiceToToCIDR(serviceInfo types.K8sServiceNamespace, endpoint types.K8sServiceEndpoint) error {
```
I would suggest to move the mutex lock here. I know you need to delete and then add the toCIDR while holding the lock but I would suggest to do that in a new function called "Replace", or something like that, so you can lock/unlock the mutex inside the repository.go file

manalibhutiyani commented on Documentation/policy.rst:
```diff
@@ -334,6 +334,50 @@ but not CIDR prefix ``10.96.0.0/12``
+Layer 3: Services
+~~~~~~~~~~~~~~~~~
+
+Services running in yor cluster can be whitelisted in Egress rules. Currently only Kubernetes services defined by their name and namespace are supported.
```
nit s/yor/your/g

nebril committed:
Respect line length, move repo locking to repo

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

tgraf commented on common/types/loadbalancer.go:
```diff
@@ -150,8 +150,8 @@ func NewLoadBalancer() *LoadBalancer {
 
 // K8sServiceNamespace is an abstraction for the k8s service + namespace types.
 type K8sServiceNamespace struct {
```
Please move this to `pkg/policy/api` if you want to make it public.

nebril replied:
Moving this will induce an import cycle between `pkg/policy/api` and `common.types` packages, so I added this type also to API, so that both `common.types.K8sServiceNamespace` and `pkg/policy/api.K8sServiceNamespace` have the same underlying type.

nebril committed:
Restrict egress ToCIDR gen to headless k8s svcs

ServiceInfo type has additional boolean field which determines whether
the service is headless or not.

This information is used to decide whether to populate/delete egress
ToCIDR rules.

Docs are updated to reflect the change.

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

tgraf commented on pkg/policy/repository.go:
```diff
@@ -369,3 +373,85 @@ func (p *Repository) NumRules() int {
 }
+
+// ConvertToK8sServiceToToCIDR traverses all egress rules and matches them against provided serviceInfo. If a matching egress rule is found it is populated with ToCIDR and ToPorts entries based on endpoint object.
+func (p *Repository) ConvertToK8sServiceToToCIDR(serviceInfo types.K8sServiceNamespace, endpoint types.K8sServiceEndpoint) error {
```
I would rather not polute `pkg/policy` with Kubernetes specific code. Instead, this should be generic functionality. In the future, this translation will happen for non Kubernetes use cases as well. What you want here is to introduce a concept of a `Translator`. It should be a go interface that `pkg/k8s` implemented. We then allow the translator to be called.

```
type Translator interface {
     TranslateRule(rule Rule)
}
```

```
func (p *Repository) RunTranslator(translator Translator) error {
    for _, rule := range p.rules {
        if err := translator.TranslateRule(rule); err != nil {
            return err
        }
    }
}
```

nebril replied:
Agreed, will do.

nebril replied:
Done

tgraf commented on pkg/policy/repository.go:
```diff
@@ -369,3 +373,85 @@ func (p *Repository) NumRules() int {
+		}
+		if !found {
+			cidr := net.IPNet{IP: epIP.Mask(mask), Mask: mask}
+			egress.ToCIDR = append(egress.ToCIDR, api.CIDR(cidr.String()))
```
What worries me a bit is that by looking at the policy rule, it is not clear what part of `ToCIDR` is specified by the user and what part has been automatically populated. I see multiple options:

 - We add another rule which contains the automatically generated values. This keeps user-specified values completely separate from the automatically generated part.
 - We add separate CIDR and Ports fields for automatically generated values and keep it in the same rule.

nebril replied:
Isn't the fact that we are only manipulating `ToCIDR` rules within egress rules which have `ToServices` defined enough?

tgraf replied:
A rule could have both, `ToCIDR` and `ToServices`. How do you know if you do `GET /policy` whether a CIDR value has been provided manually or has been populated?

tgraf replied:
I'm also fine to disallow the combination of `ToCIDR` and `ToServices` but I would prefer if we have as few ways as possible for invalid policy to be specified.

nebril replied:
I think disallowing the combination would make sense from semantic POV, but since we don't give users immediate feedback for submitting wrong policies I agree with the sentiment.

Could this be done as another PR, as I feel that this PR is already getting quite big?

tgraf replied:
I'm fine with that as long as we know and agree on the desired end state.

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -476,6 +476,12 @@ func (d *Daemon) endpointAddFn(obj interface{}) {
+
+	d.policy.Mutex.Lock()
+	defer d.policy.Mutex.Unlock()
+	if err := d.policy.ConvertToK8sServiceToToCIDR(svcns, *newSvcEP); err != nil {
```
I agree with @aanm  that the locking should not be done from the outside.

aanm reviewed (dismissed):
LGTM, please update the `vendor.conf` with the new dependencies added so we can merge it.

The remaining comments can be fixed after

nebril committed:
Move ToCIDR gen logic to k8s package

policy.Translator interface is created to allow policy repo to accept
custom policy rule manipulation logic

Signed-off-by: Maciej Kwiek <maciej@covalent.io>

aalemayhu reviewed (approved):
docs bits LGTM.

rlenglet commented on pkg/k8s/rule_translate.go:
```diff
@@ -0,0 +1,180 @@
+			return fmt.Errorf("Unable to parse ip: %s", ip)
+		}
+
+		found := false
```
I wouldn't care too much about de-duplicating CIDR IP address prefixes. @ianvernon has a change ready to de-duplicate and aggregate all CIDRs afterwards.

nebril replied:
I saw his PR after implementing that, will definitely use it.

nebril replied:
His PR is not yet merged in though, but I will add an issue to track adding the use of his code here.

rlenglet replied:
OK, LGTM.

rlenglet commented on pkg/k8s/rule_translate.go:
```diff
@@ -0,0 +1,180 @@
+		}
+		if !found {
+			cidr := net.IPNet{IP: epIP.Mask(mask), Mask: mask}
+			egress.ToCIDR = append(egress.ToCIDR, api.CIDR(cidr.String()))
```
In general, a basic principle in any REST API is that entities must never be modified except as a result of a client requests.
I.e. if I import a policy with `cilium policy import`, I should always get the exact same policy from `cilium policy get` afterwards. User input should never be modified.
You're violating that basic principle.

nebril replied:
This is true in context of using `cilium` as a user-facing client. This code is kubernetes-specific and alters what seems like an internal representation of k8s CiliumNetworkPolicy Custom Resource Definition. User which would be affected by this code would use `kubectl create -f policy.yaml`, and see the same objects when he retrieves his policy from kubernetes with `kubectl get cnp -o yaml`, as this code doesn't manipulate k8s objects.

Please let me know if this is acceptable for you.

rlenglet replied:
Got it. I didnt understand that. Thanks!

rlenglet replied:
What's preventing a user from defining a rule with a non-empty `ToServices` rule and load it with `cilium policy import`? What is that supposed not to work?
Has that been tested? I don't see any integration / component tests in this PR.

rlenglet commented on pkg/k8s/rule_translate.go:
```diff
@@ -0,0 +1,180 @@
+		}
+	}
+
+	egress.ToCIDR = newToCIDR
```
Same comment as above.
User input SHOULD never be modified.

nebril replied:
Same reply as above.

rlenglet commented on pkg/policy/policy.go:
```diff
@@ -99,3 +100,8 @@ func (s *SearchContext) String() string {
 	return strconv.Itoa(s.Depth * 2)
 }
+
+// Translator is an interface for altering policy rules
```
User input, e.g. policy rules, should never be modified.

nebril replied:
Same reply as above, should the comment here be worded in a way that should ensure that this shouldn't be used on user input?

tgraf commented on Documentation/policy.rst:
```diff
@@ -334,6 +334,53 @@ but not CIDR prefix ``10.96.0.0/12``
+~~~~~~~~~~~~~~~~~
+
+Services running in your cluster can be whitelisted in Egress rules. Currently only headless
+Kubernetes services defined by their name and namespace are supported. More documentation on HeadlessServices_.
```
Add: Future versions of Cilium will support specifying non Kubernetes services and services which are backed by pods.

nebril replied:
Done

tgraf commented on Documentation/policy.rst:
```diff
@@ -334,6 +334,53 @@ but not CIDR prefix ``10.96.0.0/12``
+                ToServices []Service `json:"toServices,omitempty"`
+                // [...]
+        }
+
```
Also add the `Service` struct here

nebril replied:
Done

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -404,18 +404,22 @@ func (d *Daemon) serviceAddFn(obj interface{}) {
 	clusterIP := net.ParseIP(svc.Spec.ClusterIP)
-	newSI := types.NewK8sServiceInfo(clusterIP)
+	headless := false
+	if svc.Spec.ClusterIP == "None" {
```
Do we need a `strings.ToLower()`?

nebril replied:
Only `None` is a valid value. `none` in clusterIP yields this error from kubectl: `spec.clusterIP: Invalid value: "none": must be empty, 'None', or a valid IP address`

tgraf replied:
I understand. There is a common Robustness Principle or Postel's law which is typically followed in such cases:
Be conservative in what you do, be liberal in what you accept from others (often reworded as "Be conservative in what you send, be liberal in what you accept").

nebril replied:
Good point, done.

rlenglet commented on pkg/k8s/rule_translate.go:
```diff
@@ -0,0 +1,180 @@
+			if err != nil {
+				return err
+			}
+			if !cidr.Contains(epIP) {
```
How do you decide whether a CIDR that you're removing is one that was specified by the user in `ToCIDR`, vs. one that you derived from `ToServices`?
Please add a test with a rule that has both a *user-specified* `ToCIDR` and a `ToServices` that result into overlapping or identical CIDR prefixes, and test that `deleteToCidrFromEndpoint` doesn't remove the `ToCIDR` that was specified by the user. And fix this method if that doesn't work.

rlenglet commented on pkg/k8s/rule_translate.go:
```diff
@@ -0,0 +1,180 @@
+
+// deleteToCidrFromEndpoint takes an egress rule and removes
+// ToCIDR rules matching endpoint
+func deleteToCidrFromEndpoint(
```
Cidr -> CIDR

rlenglet replied:
Is this method called anywhere?
It doesn't seem to work, cf. my other comment below.
If it's not used anywhere, remove it.
